tools/autograd/templates/python_variable_methods.cpp

```diff
@@ -528,29 +529,56 @@ static PyObject * THPVariable_storage_type(PyObject* self, PyObject* arg)
   END_HANDLE_TH_ERRORS
 }
 
+// Cases:
+// 1) arg matches self -> no-op
+// 2) arg is true, self doesn't requires_grad -> do the equivalent of detach() and set_requires_grad.
+// 3) arg is false, self requires_grad -> valid if self is a leaf.
+static Variable as_requires_grad(Tensor self, bool requires_grad) {
+  auto& self_ = static_cast<torch::autograd::Variable&>(self);
+  auto self_requires_grad = self_.requires_grad();
+  if (self_requires_grad == requires_grad) {
+    return self_;
+  } else if (requires_grad) {
+    auto res = make_variable(self_.data(), /*requires_grad=*/true);
+    res.set_version_counter(self_.version_counter());
+    return res;
```

Comment:
This can still have some undesirable effects if you create two independent things that `require_grad` (in-place modifications to one of them don't carry to the gradient of the other). Eh, I think we can merge this and have a big warning saying that you shouldn't do it more than once, and I'll need to think about this a bit more.

Reply:
I'm a little dubious about adding new unsafe APIs.  We could just have a separate function that is always inplace `requires_grad_()` or `requires_grad_(requires_grad=True)`.